### PR TITLE
Preformance Fixes for EmbeddingGrad OP at FP16 Accuracy

### PR DESCRIPTION
**Pull request by LUZY0726**
### PR types
Performance optimization

### PR changes
OPs

### Description
Pcard-70459
Issue:
In non-deterministic computations, the performance of EmbeddingGrad operator at fp16 precision is much lower than its performance at fp32 precision in the case of atomicAdd frequent writes to the same location. This PR aims to fix the performance of fp16 precision in such cases.

Solution:
I design an embedding backward kernel ('LaunchEmbeddingGradHalf2Float') for FP16 precision that can efficient improve the performance through use atomicAdd(float) in some cases of atomicAdd frequent writes to the same location. I use a parameter 'para' to describe the frequency of writing to the same address and set a threshold to 0.5. If 'para' is over than the threshold which indicate the frequency is high enough, so I would call function 'LaunchEmbeddingGradHalf2Float' to use atomicAdd(float) to fix performance. This PR needs alloc some extra GPU memory, it depends on the mapping relationship.

Perf tests:
test case - embedding test in benchmark (includes 4 configs for embedding backward)
unit: ms
image

### CODE REVIEW

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 1]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
   }
 }
 
+template <typename T, typename IdT>
+__global__ void ScatterFloat2Half(T* table,
+                                  float* table_used,
+                                  const IdT* unique_ids,
+                                  const int64_t D,
+                                  const int64_t unique_num) {
```
Comment (umiswing) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu:85]: table的类型如果固定是fp16的话，可以去掉模版参数中的T。看代码指针是const，可以在*后面加const修饰。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 2]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+__global__ void EmbeddingGradHalf2Float(T* table,
+                                        float* table_used,
+                                        const T* output,
+                                        const IdT* ids,
+                                        const IdT* unique_ids,
+                                        IdT* ids_map,
+                                        const int64_t N,
+                                        const int64_t K,
+                                        const int64_t D,
+                                        const int64_t unique_num) {
```
Comment (umiswing) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu:108]: table的类型如果固定是fp16的话，可以去掉模版参数中的T。看代码指针是const，可以在*后面加const修饰。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 3]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+    auto id = static_cast<int64_t>(ids_map[idy]);
+    const T* out = output + idy * D;
+    float* tab_used = table_used + id * D;
+#ifdef PADDLE_WITH_CUDA
+    for (int i = idx; i < D; i += blockDim.x) {
+      phi::CudaAtomicAdd(&tab_used[i], static_cast<float>(out[i]));
+    }
+#else
+    for (int i = idx; i < D; i += blockDim.x) {
+      phi::CudaAtomicAdd(&tab_used[i], out[i]);
```
Comment (umiswing) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 请教下，这个地方需要加static_cast吗？

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 4]
```diff
@@ -116,8 +252,19 @@ struct EmbeddingGradCUDAFunctor {
...
+            LaunchEmbeddingGradHalf2Float<T, IdT>(
+                dev_ctx_, ids, d_output, d_table, N, D, K);
+          } else {
+            EmbeddingGrad<T, IdT><<<grids, threads, 0, dev_ctx_.stream()>>>(
+                d_table, d_output, ids, N, K, D);
+          }
+        } else {
+          EmbeddingGrad<T, IdT><<<grids, threads, 0, dev_ctx_.stream()>>>(
+              d_table, d_output, ids, N, K, D);
+        }
```
Comment (umiswing) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu:275]: 这里建议通过模版特化的方式加一个dispatch，根据模版参数T的类型调不同的specialization。

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 5]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+                                   const IdT* d_ids,
+                                   const T* d_output,
+                                   T* d_table,
+                                   int64_t N,
+                                   int64_t D,
+                                   int64_t K) {
+  // Get `d_unique_ids' which contains the unique value of `d_ids'
+  // and the length of `d_unique_ids' (unique_num).
+  thrust::device_vector<IdT> d_unique_idsvec(K);
+  thrust::copy(d_ids, d_ids + K, d_unique_idsvec.begin());
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: thrust的接口需要加上stream信息，比如 `thrust::copy(thrust::cuda::par.on(dev_ctx.stream()), d_ids, d_ids + K, d_unique_idsvec.begin())`

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 6]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+      d_unique_idsvec.end());
+  const int unique_num = d_unique_idsvec.size();
+  IdT* d_unique_ids = thrust::raw_pointer_cast(d_unique_idsvec.data());
+
+  // Alloc a space of size (unique_num * D * sizeof(float)) for using
+  // atomicAdd(float).
+  float* d_table_tmp;
+  IdT* d_ids_map;
+#ifdef PADDLE_WITH_HIP
+  PADDLE_ENFORCE_GPU_SUCCESS(hipMalloc(reinterpret_cast<void**>(&d_table_tmp),
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 不建议手动malloc，可以直接用DenseTensor

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 7]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+                                       unique_num * D * sizeof(float)));
+  PADDLE_ENFORCE_GPU_SUCCESS(hipMemsetAsync(
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 可以用SetConstant

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 8]
```diff
@@ -65,6 +76,131 @@ __global__ void EmbeddingGrad(T* table,
...
+                                        unique_num * D * sizeof(float)));
+  PADDLE_ENFORCE_GPU_SUCCESS(cudaMemsetAsync(
+      d_table_tmp, 0, unique_num * D * sizeof(float), ctx.stream()));
+  PADDLE_ENFORCE_GPU_SUCCESS(
+      cudaMalloc(reinterpret_cast<void**>(&d_ids_map), K * sizeof(IdT)));
+#endif
+
+  // Get mapping array (`d_ids_map'), which records the mapping relationship
+  // from `d_ids' to `d_unique_ids' and calculate with atomicAdd(float).
+  const int gridx = 2 * ctx.GetSMCount();
```
Comment (zkh2016) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 这个应该根据数据规模来设置？

Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 7 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 6 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 5 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 3 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 需要，已修正

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 2 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu:108]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 1 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu:85]: done

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu (hunk 4 above)
Comment (LUZY0726) [reply paddle/phi/kernels/gpu/embedding_grad_kernel.cu:275]: 好的

**Review by zkh2016 — DISMISSED** [review]

**Pull request closed without merging**